### PR TITLE
Task-55618: Implement update of mail integration setting

### PR DESCRIPTION
**Pull request by hassenelkamel-exo**
After this change, it will be possible to update mail integration setting.

### CODE REVIEW

File: mail-integration-services/src/test/java/org/exoplatform/mailintegration/dao/MailIntegrationDaoTest.java [hunk 1]
```diff
@@ -95,6 +95,27 @@ public void testGetMailIntegrationSettingsByUserId() {
...
   }
 
+  public void testUpdateMailIntegrationSetting() {
+    //Given
+    MailIntegrationSettingEntity createdMailIntegrationSettingEntity = createMailIntegrationSettingEntity();
+    createdMailIntegrationSettingEntity = mailIntegrationDAO.create(createdMailIntegrationSettingEntity);
+
+    // When
+    createdMailIntegrationSettingEntity.setAccount("updatedAccount");
+    MailIntegrationSettingEntity updatedMailIntegration = mailIntegrationDAO.update(createdMailIntegrationSettingEntity);
```
Comment (azayati) [inline mail-integration-services/src/test/java/org/exoplatform/mailintegration/dao/MailIntegrationDaoTest.java]:
```suggestion
    MailIntegrationSettingEntity updatedMailIntegrationSetting = mailIntegrationDAO.update(createdMailIntegrationSettingEntity);
```

File: mail-integration-services/src/test/java/org/exoplatform/mailintegration/dao/MailIntegrationDaoTest.java [hunk 2]
```diff
@@ -95,6 +95,27 @@ public void testGetMailIntegrationSettingsByUserId() {
 
   }
 
+  public void testUpdateMailIntegrationSetting() {
+    //Given
+    MailIntegrationSettingEntity createdMailIntegrationSettingEntity = createMailIntegrationSettingEntity();
+    createdMailIntegrationSettingEntity = mailIntegrationDAO.create(createdMailIntegrationSettingEntity);
+
+    // When
+    createdMailIntegrationSettingEntity.setAccount("updatedAccount");
```
Comment (azayati) [inline mail-integration-services/src/test/java/org/exoplatform/mailintegration/dao/MailIntegrationDaoTest.java]: put this ion the Given section

File: mail-integration-services/src/test/java/org/exoplatform/mailintegration/storage/MailIntegrationStorageTest.java [hunk 3]
```diff
@@ -101,6 +101,30 @@ public void testGetMailIntegrationSettings() throws Exception { // NOSONAR
     assertNotNull(retrievedMailIntegrationSettings);
     assertEquals(mailIntegrationSettingEntities.size(), retrievedMailIntegrationSettings.size());
   }
+
+  @PrepareForTest({ EntityMapper.class })
+  @Test
+  public void testUpdatedMailIntegrationSetting() throws Exception { // NOSONAR
```
Comment (azayati) [inline mail-integration-services/src/test/java/org/exoplatform/mailintegration/storage/MailIntegrationStorageTest.java]:
```suggestion
  public void testUpdateMailIntegrationSetting() throws Exception { // NOSONAR
```

File: mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java [hunk 4]
```diff
@@ -190,4 +184,36 @@ public Response getMessageById(@ApiParam(value = "Message id", required = true)
...
+  @ApiResponses(
+          value = { @ApiResponse(code = HTTPStatus.NO_CONTENT, message = "Request fulfilled"),
+                  @ApiResponse(code = HTTPStatus.NOT_FOUND, message = "Object not found"),
+                  @ApiResponse(code = HTTPStatus.BAD_REQUEST, message = "Invalid query input"),
+                  @ApiResponse(code = HTTPStatus.UNAUTHORIZED, message = "Unauthorized operation"),
+                  @ApiResponse(code = HTTPStatus.INTERNAL_ERROR, message = "Internal server error"),
+          }
+  )
+  public Response updateMailIntegrationSetting(
+                                               @ApiParam(value = "Event object to update", required = true)
```
Comment (azayati) [inline mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java]:
```suggestion
                                               @ApiParam(value = "Mail integration setting object to update", required = true)
```

File: mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java [hunk 5]
```diff
@@ -190,4 +184,36 @@ public Response getMessageById(@ApiParam(value = "Message id", required = true)
...
+                                               MailIntegrationSettingRestEntity mailIntegrationSettingEntity) {
+    if (mailIntegrationSettingEntity == null) {
+      return Response.status(Response.Status.BAD_REQUEST).build();
+    }
+    long userIdentityId = MailIntegrationUtils.getCurrentUserIdentityId(identityManager);
+    try {
+      MailIntegrationSetting mailIntegrationSetting = RestEntityBuilder.toMailIntegrationSetting(mailIntegrationSettingEntity,
+                                                                                                 userIdentityId);
+      MailIntegrationSetting createdMailIntegrationSetting =
```
Comment (azayati) [inline mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java]:
```suggestion
      MailIntegrationSetting updatedMailIntegrationSetting =
```

File: mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java [hunk 6]
```diff
@@ -190,4 +184,36 @@ public Response getMessageById(@ApiParam(value = "Message id", required = true)
...
+                                                           mailIntegrationService.updateMailIntegrationSetting(mailIntegrationSetting);
+      return Response.ok(createdMailIntegrationSetting).build();
+    } catch (Exception e) {
+      LOG.error("Error when creating mail integration setting ", e);
```
Comment (azayati) [inline mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java]:
```suggestion
      LOG.error("Error when updating mail integration setting ", e);
```

File: mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java [hunk 7]
```diff
@@ -190,4 +184,36 @@ public Response getMessageById(@ApiParam(value = "Message id", required = true)
...
+      return Response.status(Response.Status.BAD_REQUEST).build();
+    }
+    long userIdentityId = MailIntegrationUtils.getCurrentUserIdentityId(identityManager);
+    try {
+      MailIntegrationSetting mailIntegrationSetting = RestEntityBuilder.toMailIntegrationSetting(mailIntegrationSettingEntity,
+                                                                                                 userIdentityId);
+      MailIntegrationSetting createdMailIntegrationSetting =
+                                                           mailIntegrationService.updateMailIntegrationSetting(mailIntegrationSetting);
+      return Response.ok(createdMailIntegrationSetting).build();
+    } catch (Exception e) {
```
Comment (azayati) [inline mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java:217]: you should add IllegalAccessException when update is not authorized

File: mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java [hunk 8]
```diff
@@ -190,4 +184,36 @@ public Response getMessageById(@ApiParam(value = "Message id", required = true)
...
     }
   }
+
+  @PUT
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  @RolesAllowed("users")
+  @ApiOperation(value = "Update an existing mail integration setting", httpMethod = "PUT", response = Response.class, consumes = "application/json")
+  @ApiResponses(
+          value = { @ApiResponse(code = HTTPStatus.NO_CONTENT, message = "Request fulfilled"),
```
Comment (azayati) [inline mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java]: not needed

File: mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java [hunk 9]
```diff
@@ -190,4 +184,36 @@ public Response getMessageById(@ApiParam(value = "Message id", required = true)
...
+                  @ApiResponse(code = HTTPStatus.NOT_FOUND, message = "Object not found"),
```
Comment (azayati) [inline mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java]: a check on mailIntegrationSetting existance is needed as for deleteMailIntegrationSetting

File: mail-integration-api/src/main/java/org/exoplatform/mailintegration/service/MailIntegrationService.java [hunk 10]
```diff
@@ -82,4 +82,11 @@ public interface MailIntegrationService {
   MessageRestEntity getMessageById(long mailIntegrationSettingId,
                                    String messageId,
                                    long userIdentityId) throws IllegalAccessException;
+
+  /**
+   *
```
Comment (azayati) [inline mail-integration-api/src/main/java/org/exoplatform/mailintegration/service/MailIntegrationService.java:88]: Add description here

File: mail-integration-api/src/main/java/org/exoplatform/mailintegration/service/MailIntegrationService.java [hunk 11]
```diff
@@ -82,4 +82,11 @@ public interface MailIntegrationService {
...
+   * @param mailIntegrationSetting {@link MailIntegrationSetting} object to update
+   * @return updated an {@link MailIntegrationSetting} object
```
Comment (azayati) [inline mail-integration-api/src/main/java/org/exoplatform/mailintegration/service/MailIntegrationService.java]:
```suggestion
   * @return updated {@link MailIntegrationSetting} object
```

**Review by azayati — CHANGES_REQUESTED** [review]
see review

File: mail-integration-api/src/main/java/org/exoplatform/mailintegration/service/MailIntegrationService.java [hunk 12]
```diff
@@ -84,9 +84,10 @@ MessageRestEntity getMessageById(long mailIntegrationSettingId,
...
 
   /**
+   * Update mail integration setting
    *
    * @param mailIntegrationSetting {@link MailIntegrationSetting} object to update
-   * @return updated an {@link MailIntegrationSetting} object
+   * @return updated {@link MailIntegrationSetting} object
    */
-  MailIntegrationSetting updateMailIntegrationSetting(MailIntegrationSetting mailIntegrationSetting) throws IllegalAccessException;
+  MailIntegrationSetting updateMailIntegrationSetting(MailIntegrationSetting mailIntegrationSetting);
```
Comment (azayati) [inline mail-integration-api/src/main/java/org/exoplatform/mailintegration/service/MailIntegrationService.java]: keep the IllegalAccessException to be catched when the user is not authorized to update

**Review by azayati — CHANGES_REQUESTED** [review]
see review

File: mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java (hunk 9 above)
Comment (hassenelkamel-exo) [reply mail-integration-services/src/main/java/org/exoplatform/mailintegration/rest/MailIntegrationRest.java]: it had been done

**Review by azayati — DISMISSED** [review]
LGTM after QA validation

**Review by azayati — APPROVED** [review]
LGTM

**Pull request merged**